File: BasicSample/app/src/main/java/com/example/android/persistence/ui/ProductAdapter.java
/*
 * Copyright 2017, The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.example.android.persistence.ui;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;

import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import com.example.android.persistence.databinding.ProductItemBinding;
import com.example.android.persistence.model.Product;
import com.example.android.persistence.R;

import java.util.List;

public class ProductAdapter extends RecyclerView.Adapter<ProductAdapter.ProductViewHolder> {
    //产品 list
    List<? extends Product> mProductList;

    @Nullable//点击回调
    private final ProductClickCallback mProductClickCallback;

    public ProductAdapter(@Nullable ProductClickCallback clickCallback) {
        mProductClickCallback = clickCallback;
        //notifyDataSetChanged时导致图片闪烁 需要重写 getItemId 返回唯一id.
        setHasStableIds(true);
    }
    //显示产品list
    public void setProductList(final List<? extends Product> productList) {
        if (mProductList == null) {
            mProductList = productList;
            //从位置0开始插入
            notifyItemRangeInserted(0, productList.size());
        } else {
            //计算数据集差异
            DiffUtil.DiffResult result = DiffUtil.calculateDiff(new DiffUtil.Callback() {
                @Override
                public int getOldListSize() {
                    return mProductList.size();
                }

                @Override
                public int getNewListSize() {
                    return productList.size();
                }

                @Override
                public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
                    return mProductList.get(oldItemPosition).getId() ==
                            productList.get(newItemPosition).getId();
                }

                @Override
                public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
                    Product newProduct = productList.get(newItemPosition);
                    Product oldProduct = mProductList.get(oldItemPosition);
                    return newProduct.getId() == oldProduct.getId()
                            && TextUtils.equals(newProduct.getDescription(), oldProduct.getDescription())
                            && TextUtils.equals(newProduct.getName(), oldProduct.getName())
                            && newProduct.getPrice() == oldProduct.getPrice();
                }
            });
            mProductList = productList;
            result.dispatchUpdatesTo(this);
        }
    }

    @Override
    @NonNull
    public ProductViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ProductItemBinding binding = DataBindingUtil
                .inflate(LayoutInflater.from(parent.getContext()), R.layout.product_item,
                        parent, false);
        binding.setCallback(mProductClickCallback);
        return new ProductViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull ProductViewHolder holder, int position) {
        holder.binding.setProduct(mProductList.get(position));
        holder.binding.executePendingBindings();
    }

    @Override
    public int getItemCount() {
        return mProductList == null ? 0 : mProductList.size();
    }

    @Override
    public long getItemId(int position) {
        return mProductList.get(position).getId();
    }

    static class ProductViewHolder extends RecyclerView.ViewHolder {

        final ProductItemBinding binding;

        public ProductViewHolder(ProductItemBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }
}

File: BasicSample-kt/app/src/main/java/com/example/kodomo/persistence/ui/ProductAdapter.kt
package com.example.kodomo.persistence.ui

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.example.kodomo.persistence.R
import com.example.kodomo.persistence.databinding.ProductItemBinding
import com.example.kodomo.persistence.model.Product
import java.util.*


/**
 * @author linxiaotao
 * 2018/12/24 下午5:33
 */
class ProductAdapter(val clickCallback: ProductClickCallback) : RecyclerView.Adapter<ProductAdapter.ProductViewHolder>() {

    init {
        setHasStableIds(true)
    }

    val mProductList = mutableListOf<Product>()

    fun setProductList(productList: List<Product>) {
        if (mProductList.isEmpty()) {
            mProductList.addAll(productList)
            notifyItemRangeInserted(0, productList.size)
        } else {
            val result = DiffUtil.calculateDiff(object : DiffUtil.Callback() {
                override fun getOldListSize(): Int {
                    return mProductList.size
                }

                override fun getNewListSize(): Int {
                    return productList.size
                }

                override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                    return mProductList[oldItemPosition].id() ==
                            productList[newItemPosition].id()
                }

                override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                    return productList[newItemPosition].id() == mProductList[oldItemPosition].id()
                            && Objects.equals(productList[newItemPosition].description(), mProductList[oldItemPosition].description())
                            && Objects.equals(productList[newItemPosition].name(), mProductList[oldItemPosition].name())
                            && productList[newItemPosition].price() == mProductList[oldItemPosition].price()
                }
            })

            mProductList.clear()
            mProductList.addAll(productList)
            result.dispatchUpdatesTo(this)
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ProductViewHolder {
        val binding: ProductItemBinding = DataBindingUtil.inflate(
                LayoutInflater.from(parent.context),
                R.layout.product_item,
                parent,
                false
        )
        binding.callback = clickCallback
        return ProductViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ProductViewHolder, position: Int) {
        holder.binding.product = mProductList[position]
        holder.binding.executePendingBindings()
    }

    override fun getItemCount(): Int {
        return mProductList.size
    }

    override fun getItemId(position: Int): Long {
        return mProductList[position].id().toLong()
    }

    class ProductViewHolder(val binding: ProductItemBinding) : RecyclerView.ViewHolder(binding.root) {


    }

}